src/server/services/DripRequestHandler.spec.ts

```diff
@@ -0,0 +1,149 @@
+    storageFileName = `./test-storage.db`;
+    storage = new ActionStorage(storageFileName);
+    await storage.isValid({ addr: "anyone, effectively awaiting sqlite initialization." });
+    handler = new DripRequestHandler(actionsMock, storage, recaptcha);
+  });
+
+  afterEach(async () => {
+    await fs.rmSync(storageFileName);
+  });
+
+  /**
+   * Normal operation, the requests are coming from the matrix bot in a sibling docker container.
+   */
+  describe("Without external access", () => {
+    const defaultRequest = {
+      external: false,
+      amount: "0.5",
+      parachain_id: "1002",
+      address: "123",
+      sender: "someone",
+    } as const;
+
+    it("Goes through one time", async () => {
+      {
+        const result = await handler.handleRequest(defaultRequest);
+        expect(result).toEqual({ hash: "0x123" });
+      }
+      {
+        const result = await handler.handleRequest(defaultRequest);
+        expect("error" in result).toBeTruthy();
+      }
+    });
+
+    it("Returns an error response", async () => {
+      const result = await handler.handleRequest({ ...defaultRequest, address: "unlucky" });
+      expect(result).toEqual({ error: "An error occurred when sending tokens" });
+    });
+
+    it("Doesn't allow a repeated address", async () => {
+      await storage.saveData({ addr: defaultRequest.address });
+      const result = await handler.handleRequest(defaultRequest);
+      expect(result).toEqual({ error: "Requester has reached their daily quota. Only request once per day." });
+    });
+
+    it("Doesn't allow a repeated user", async () => {
+      await storage.saveData({ username: defaultRequest.sender, addr: "other" });
+      const result = await handler.handleRequest(defaultRequest);
+      expect(result).toEqual({ error: "Requester has reached their daily quota. Only request once per day." });
+    });
+
+    it("Doesn't allow a rich address user", async () => {
+      const result = await handler.handleRequest({ ...defaultRequest, address: "rich" });
+      expect(result).toEqual({ error: "Requester's balance is over the faucet's balance cap" });
+    });
+
+    it("Parity members are privileged in terms of repeated requests", async () => {
+      await storage.saveData({ username: "someone:parity.io", addr: defaultRequest.address });
+      const result = await handler.handleRequest({ ...defaultRequest, sender: "someone:parity.io" });
+      expect(result).toEqual({ hash: "0x123" });
+    });
+
+    it("Parity members are privileged in terms of balance cap", async () => {
+      const result = await handler.handleRequest({ ...defaultRequest, sender: "someone:parity.io", address: "rich" });
+      expect(result).toEqual({ hash: "0x123" });
+    });
+  });
+
+  /**
+   * The requests are coming from external requesters.
+   */
+  describe("With external access", () => {
+    const defaultRequest = {
+      external: true,
+      amount: "0.5",
+      parachain_id: "1002",
```

Comment:
tests are great!  however, since parachain_id is optional, should be cases with & without parachain id

Reply:
Added those cases.